torch/nn/parallel/distributed.py

```diff
@@ -673,6 +673,9 @@ def no_sync(self):
             self.require_backward_grad_sync = old_require_backward_grad_sync
 
     def forward(self, *inputs, **kwargs):
+        if torch.is_grad_enabled() and self.require_backward_grad_sync:
+            self.logger.set_runtime_stats()
```

Comment:
Have we profiled this? It might be expensive to compute stats for every iteration. We might want to do some sampling here.

Reply:
yes, I profiled. There is no regression for both resnet and bert models. I think the main reason is: 1) event is initialized only once (event initialization is expensive). 2) event synchronization happens after optimizer step is done. so synchronization is cheap here. 3) event recording and average operation is actually cheap here. 

but let me think about whether we want to add sampling for safety and flexibility.

Reply:
by the way, although we collect stats at every iteration, we only log data to downstream stable at 3 selected iterations. so the time spending on logging to downstream table is also trivial